src/journal/journald-native.c

```diff
@@ -193,14 +195,8 @@ void server_process_native_message(
 
                                 } else if (l >= 8 &&
                                            startswith(p, "MESSAGE=")) {
-                                        char *t;
-
-                                        t = strndup(p + 8, l - 8);
-                                        if (t) {
-                                                free(message);
-                                                message = t;
```

Comment:
Good catch, thanks!

I figure I will revert to my original solution. I will ```strndup()```, that will cut off everything after first NUL byte, and ```message_len``` will be  equal to ```strlen()``` of newly allocated string.

After all I think it is legitimate for us to not accept NUL bytes in native messages. Only way how to preserve log messages 100% verbatim would be ```/dev/log -> syslog.socket```.